eventkitd/utils/resumablecompressed/reader.go

```diff
@@ -0,0 +1,63 @@
+package resumablecompressed
+
+import (
+	"compress/zlib"
+	"errors"
+	"io"
+
+	"github.com/jtolio/eventkit/eventkitd/utils/delimited"
+)
+
+type Reader struct {
+	r       *delimited.Reader
+	current io.ReadCloser
+}
+
+func NewReader(base io.Reader) *Reader {
+	r := delimited.NewReader(base)
+	return &Reader{
+		r: r,
+	}
+}
+
+type readerFunc func(p []byte) (n int, err error)
+
+func (f readerFunc) Read(p []byte) (n int, err error) { return f(p) }
+
+func (r *Reader) Read(p []byte) (n int, err error) {
+	if r.current == nil {
+		for r.r.Delimited() {
+			r.r.Advance()
```

Comment:
Note, if `r.r.` can provide information about whether there's something to read at all, you can rid of the wrapper used inside zlib.NewReader.

Reply:
so, unfortunately, the format has some bad possibilities. in particular, there may be arbitrarily many data frames of zero bytes long. each frame at minimum has a header, so we don't know how long we have to read to get to the next useful frame. i suppose we could contrive the format so that zero byte data frames are forbidden and then we would be able to know more about what comes next with one-frame-header-lookahead, but as it stands we actually don't know what comes next, or if there's anything useful to read at all.  the current expectation of delimited.Reader is that between delimiters, users will read until io.EOF, before checking if Delimited() is true. maybe this is a hint that the ux of delimited.Reader is still not great, especially since i'm not really doing that correctly here.